plugins/Login/Controller.php

```diff
@@ -303,14 +303,15 @@ protected function authenticateAndRedirect($login, $password, $urlToRedirect = f
         $this->passwordResetter->removePasswordResetInfo($login);
 
         if (empty($urlToRedirect)) {
-            $referrer = Url::getReferrer();
-            $module = Common::getRequestVar('module', '', 'string');
+            $redirect = Common::unsanitizeInputValue(Common::getRequestVar('form_redirect', false));
+            $redirectParams = UrlHelper::getArrayFromQueryString(UrlHelper::getQueryFromUrl($redirect));
+            $module = Common::getRequestVar('module', '', 'string', $redirectParams);
             // when module is login, we redirect to home...
-            if ($module !== 'Login' && $module !== Piwik::getLoginPluginName() && $referrer) {
-                $host = Url::getHostFromUrl($referrer);
+            if ($module !== 'Login' && $module !== Piwik::getLoginPluginName() && $redirect) {
+                $host = Url::getHostFromUrl($redirect);
                 // we only redirect to a trusted host
                 if ($host && Url::isValidHost($host)) {
```

Comment:
I just realised I should have maybe adjusted the description for the config setting

```
; Whether to enable trusted host checking. This can be disabled if you're running Matomo
; on several URLs and do not wish to constantly edit the trusted host list.
enable_trusted_host_check = 1
```

When disabled, you can redirect any user to any url... and now that it's even possible by GET parameter and not referrer it may be even more important. Basically, when the setting is disabled, you can link to `https://demo.matomo.org?...&redirect=https://demo.maomo.org` for example and after the succeeded login, the user would be redirected to that page. This could be some security issue if you ask the user on that similar domain to enter the password again. Wonder how we could mention it in the config? In general we should not recommend users to disable the trusted host check.

Reply:
What if we also check that the host in the URL is the same as the Host header? Then the only way to force a redirect to a different domain, would be to send a curl request w/ the changed Host header.

Reply:
sounds good